src/core/pointcloud/qgspointcloudstatscalculator.cpp

```diff
@@ -209,13 +211,18 @@ bool QgsPointCloudStatsCalculator::calculateStats( QgsFeedback *feedback, const
 
   connect( feedback, &QgsFeedback::canceled, &mFutureWatcher, &QFutureWatcher<QgsPointCloudStatistics>::cancel );
 
-  mFutureWatcher.waitForFinished();
+  mEventLoop.reset( new QEventLoop );
 
-  if ( mFutureWatcher.isCanceled() )
-  {
-    return false;
-  }
+  connect( &mFutureWatcher, &QFutureWatcher<QgsPointCloudStatistics>::finished, this, &QgsPointCloudStatsCalculator::statisticsCalculationQtConcurrentCallFinished );
+  connect( &mFutureWatcher, &QFutureWatcher<QgsPointCloudStatistics>::canceled, mEventLoop.get(), &QEventLoop::quit );
 
+  mEventLoop->exec();
```

Comment:
An event loop is potentially unsafe to use here -- you need a guard around this to ensure that it's never used on the main thread, and only ever on a background thread.

Reply:
This statistics calculator is supposed to be executed in QgsPointCloudStatsCalculationTask , so technically it is not executed in the main thread except for tests or maybe some script that wants to use it directly.

Reply:
I managed to find the proper way to remove the event loop and get rid of concurrency problems :)